src/block-cache.ts

```diff
@@ -0,0 +1,193 @@
+    if (payload.method && ['eth_getTransactionByHash', 'eth_getTransactionReceipt'].includes(payload.method)) {
+      if (!result || !result.blockHash || result.blockHash === '0x0000000000000000000000000000000000000000000000000000000000000000') {
+        return false;
+      }
+    }
+    // otherwise true
+    return true;
+  }
+
+  // removes all block caches with block number lower than `oldBlockHex`
+  clearBefore(oldBlockHex: string): void {
+    const oldBlockNumber: number = Number.parseInt(oldBlockHex, 16);
+    // clear old caches
+    Object.keys(this.cache)
+      .map(Number)
+      .filter((num) => num < oldBlockNumber)
+      .forEach((num) => delete this.cache[num]);
+  }
+}
+
+function createBlockCacheMiddleware(
+  { blockTracker }: BlockCacheMiddlewareOptions = {},
+): JsonRpcMiddleware<string[], Block> {
+  // validate options
+  if (!blockTracker) {
+    throw new Error('createBlockCacheMiddleware - No BlockTracker specified');
+  }
+
+  // create caching strategies
+  const blockCache: BlockCacheStrategy = new BlockCacheStrategy();
+  const strategies: Record<string, BlockCacheStrategy> = {
+    perma: blockCache,
+    block: blockCache,
+    fork: blockCache,
+  };
+
+  return createAsyncMiddleware(async (req, res, next) => {
+    // allow cach to be skipped if so specified
+    if ((req as JsonRPCRequestToCache).skipCache) {
+      return next();
+    }
+    // check type and matching strategy
+    const type: string = cacheTypeForPayload(req);
+    const strategy: BlockCacheStrategy = strategies[type];
+    // If there's no strategy in place, pass it down the chain.
+    if (!strategy) {
+      return next();
+    }
+    // If the strategy can't cache this request, ignore it.
+    if (!strategy.canCacheRequest(req)) {
+      return next();
+    }
+
+    // get block reference (number or keyword)
+    let blockTag: string|undefined = blockTagForPayload(req);
+    if (blockTag === undefined) {
+      blockTag = 'latest';
+    }
+
+    // get exact block number
+    let requestedBlockNumber: string;
+    if (blockTag === 'earliest') {
+      // this just exists for symmetry with "latest"
+      requestedBlockNumber = '0x00';
+    } else if (blockTag === 'latest') {
+      // fetch latest block number
+      const latestBlockNumber = await blockTracker.getLatestBlock();
+      // clear all cache before latest block
+      blockCache.clearBefore(latestBlockNumber);
+      requestedBlockNumber = latestBlockNumber;
+    } else {
+      // We have a hex number
+      requestedBlockNumber = blockTag;
+    }
+    // end on a hit, continue on a miss
+    const cacheResult: Block|undefined = await strategy.get(req, requestedBlockNumber);
+    if (cacheResult === undefined) {
+      // cache miss
+      // wait for other middleware to handle request
+      // eslint-disable-next-line node/callback-return
+      await next();
+
+      if (res.result === undefined) {
+        return undefined;
+      }
```

Comment:
To avoid typecasting of `res.result` on `await strategy.set(req, requestedBlockNumber, res.result);`
I couldn't decide which is the better approach!!

Reply:
In this case we should cast, because `BlockCacheStrategy.set` calls `canCacheResult`, which checks if `res.result` is one of the "empty" values. This package otherwise assumes that the RPC endpoints return correct data, so there are no more runtime type checks to perform.

I made the change in 3a4c94f, to decrease turnaround.